### PR TITLE
Add support for prop_add_list()

### LINKED ISSUE #8675: proposal: add prop_add_list() to enable adding multiple text properties per call
**Is your feature request about something that is currently impossible or hard to do? Please describe the problem.**

As a part of [govim](github.com/govim/govim), a Go development plugin for vim, I'm adding support for the LSP feature ["semantic tokens"](https://microsoft.github.io/language-server-protocol/specification#textDocument_semanticTokens).

In essence it lets an external process (language server) provide syntax highlighting that is more detailed than the one provided by vim (since it is has more knowledge about the specific language in detail).

I'm using text properties to apply these highlights to a buffer, by defining a couple of different text property types (like `comment`, `namespace` or `keyword`; as defined by the language server protocol).

Since a file often has many different highlights, this yields many calls to `prop_add()`, even if we only highlight the visible part of a buffer (i.e. lines visible in a window).

**Describe the solution you'd like**
I'd like to be able to add multiple text properties (of a specific type?) in a single call.

So that I can, instead of calling (real life example only rendering the visible part of a buffer):
```
prop_add(1,1,{"type":"comment","id":2,"end_lnum":1,"end_col":55,"bufnr":1})
prop_add(2,1,{"type":"comment","id":2,"end_lnum":2,"end_col":54,"bufnr":1})
prop_add(3,1,{"type":"comment","id":2,"end_lnum":3,"end_col":50,"bufnr":1})
prop_add(5,1,{"type":"keyword","id":2,"end_lnum":5,"end_col":8,"bufnr":1})
prop_add(5,9,{"type":"namespace","id":2,"end_lnum":5,"end_col":12,"bufnr":1})
prop_add(7,1,{"type":"keyword","id":2,"end_lnum":7,"end_col":7,"bufnr":1})
prop_add(8,3,{"type":"namespace","id":2,"end_lnum":8,"end_col":8,"bufnr":1})
prop_add(9,3,{"type":"namespace","id":2,"end_lnum":9,"end_col":10,"bufnr":1})
prop_add(10,3,{"type":"namespace","id":2,"end_lnum":10,"end_col":6,"bufnr":1})
prop_add(11,6,{"type":"namespace","id":2,"end_lnum":11,"end_col":9,"bufnr":1})
prop_add(12,6,{"type":"namespace","id":2,"end_lnum":12,"end_col":11,"bufnr":1})
prop_add(13,6,{"type":"namespace","id":2,"end_lnum":13,"end_col":11,"bufnr":1})
prop_add(14,8,{"type":"namespace","id":2,"end_lnum":14,"end_col":16,"bufnr":1})
prop_add(15,3,{"type":"namespace","id":2,"end_lnum":15,"end_col":7,"bufnr":1})
prop_add(16,3,{"type":"namespace","id":2,"end_lnum":16,"end_col":10,"bufnr":1})
prop_add(17,3,{"type":"namespace","id":2,"end_lnum":17,"end_col":7,"bufnr":1})
prop_add(19,31,{"type":"namespace","id":2,"end_lnum":19,"end_col":36,"bufnr":1})
prop_add(20,35,{"type":"namespace","id":2,"end_lnum":20,"end_col":43,"bufnr":1})
prop_add(21,35,{"type":"namespace","id":2,"end_lnum":21,"end_col":41,"bufnr":1})
prop_add(22,35,{"type":"namespace","id":2,"end_lnum":22,"end_col":43,"bufnr":1})
prop_add(23,2,{"type":"namespace","id":2,"end_lnum":23,"end_col":8,"bufnr":1})
prop_add(26,1,{"type":"comment","id":2,"end_lnum":26,"end_col":84,"bufnr":1})
prop_add(27,1,{"type":"comment","id":2,"end_lnum":27,"end_col":84,"bufnr":1})
prop_add(28,1,{"type":"comment","id":2,"end_lnum":28,"end_col":64,"bufnr":1})
prop_add(30,1,{"type":"comment","id":2,"end_lnum":30,"end_col":55,"bufnr":1})
prop_add(31,1,{"type":"keyword","id":2,"end_lnum":31,"end_col":6,"bufnr":1})
prop_add(31,7,{"type":"variable","id":2,"end_lnum":31,"end_col":22,"bufnr":1})
prop_add(31,23,{"type":"type","id":2,"end_lnum":31,"end_col":26,"bufnr":1})
prop_add(31,29,{"type":"number","id":2,"end_lnum":31,"end_col":35,"bufnr":1})
prop_add(33,1,{"type":"keyword","id":2,"end_lnum":33,"end_col":5,"bufnr":1})
prop_add(33,7,{"type":"variable","id":2,"end_lnum":33,"end_col":8,"bufnr":1})
prop_add(33,9,{"type":"operator","id":2,"end_lnum":33,"end_col":10,"bufnr":1})
prop_add(33,10,{"type":"type","id":2,"end_lnum":33,"end_col":16,"bufnr":1})
prop_add(33,18,{"type":"member","id":2,"end_lnum":33,"end_col":36,"bufnr":1})
prop_add(33,37,{"type":"parameter","id":2,"end_lnum":33,"end_col":40,"bufnr":1})
prop_add(33,41,{"type":"namespace","id":2,"end_lnum":33,"end_col":48,"bufnr":1})
prop_add(33,49,{"type":"type","id":2,"end_lnum":33,"end_col":56,"bufnr":1})
prop_add(33,58,{"type":"parameter","id":2,"end_lnum":33,"end_col":59,"bufnr":1})
prop_add(33,60,{"type":"operator","id":2,"end_lnum":33,"end_col":61,"bufnr":1})
prop_add(33,61,{"type":"namespace","id":2,"end_lnum":33,"end_col":69,"bufnr":1})
prop_add(33,70,{"type":"type","id":2,"end_lnum":33,"end_col":90,"bufnr":1})
prop_add(33,93,{"type":"operator","id":2,"end_lnum":33,"end_col":94,"bufnr":1})
prop_add(33,94,{"type":"namespace","id":2,"end_lnum":33,"end_col":102,"bufnr":1})
prop_add(33,103,{"type":"type","id":2,"end_lnum":33,"end_col":117,"bufnr":1})
prop_add(33,119,{"type":"type","id":2,"end_lnum":33,"end_col":124,"bufnr":1})
prop_add(34,2,{"type":"variable","id":2,"end_lnum":34,"end_col":5,"bufnr":1})
prop_add(34,7,{"type":"variable","id":2,"end_lnum":34,"end_col":10,"bufnr":1})
prop_add(34,11,{"type":"operator","id":2,"end_lnum":34,"end_col":13,"bufnr":1})
prop_add(34,14,{"type":"variable","id":2,"end_lnum":34,"end_col":15,"bufnr":1})
prop_add(34,16,{"type":"function","id":2,"end_lnum":34,"end_col":37,"bufnr":1})
prop_add(34,38,{"type":"variable","id":2,"end_lnum":34,"end_col":41,"bufnr":1})
prop_add(34,43,{"type":"variable","id":2,"end_lnum":34,"end_col":44,"bufnr":1})
prop_add(34,45,{"type":"variable","id":2,"end_lnum":34,"end_col":57,"bufnr":1})
prop_add(34,59,{"type":"variable","id":2,"end_lnum":34,"end_col":62,"bufnr":1})
```

I could call something like (just an example, might be better ways to design the parameters) :
```
prop_add_list({"type":"command":"id":2,"bufnr":1}, [1,1,1,55], [2,1,2,54], [3,1,3,50], [26,1,26,84], [27,1,27,84], [28,1,28,64], [30,1,30,55])
prop_add_list({"type":"keyword":"id":2,"bufnr":1}, [5,1,5,8], [7,1,7,7], [31,1,31,6], [33,1,33,5])
prop_add_list({"type":"namespace":"id":2,"bufnr":1}, [5,9,5,12], [8,3,8,8], [9,3,9,10], [10,3,10,6], [11,6,11,9], [12,6,12,11], [13,6,13,11], [14,8,14,16], [15,3,15,7], [16,3,16,10], [17,3,17,7], [19,31,19,36], [20,35,20,43], [21,35,21,41], [22,35,22,43], [23,2,23,8], [33,41,33,48], [33,61,33,69], [33,94,33,102])
prop_add_list({"type":"variable":"id":2,"bufnr":1}, [31,7,31,22], [33,7,33,8], [34,2,34,5], [34,7,34,10], [34,14,34,15], [34,38,34,41], [34,43,34,44], [34,45,34,57], [34,59,34,62])
prop_add_list({"type":"type":"id":2,"bufnr":1}, [31,23,31,26], [33,10,33,16], [33,49,33,56], [33,70,33,90], [33,103,33,117], [33,119,33,124])
prop_add_list({"type":"number":"id":2,"bufnr":1}, [31,29,31,35])
prop_add_list({"type":"operator":"id":2,"bufnr":1}, [33,9,33,10], [33,60,33,61], [33,93,33,94], [34,11,34,13])
prop_add_list({"type":"member":"id":2,"bufnr":1}, [33,18,33,36])
prop_add_list({"type":"parameter":"id":2,"bufnr":1}, [33,37,33,40], [33,58,33,59])
prop_add_list({"type":"function":"id":2,"bufnr":1}, [34,16,34,37])
```

The nature of syntax highlighting requires a lot of text properties added (and removed if you only want to show the visible lines).

**Describe alternatives you've considered**
None, I currently call `prop_add` multiple times each time the highlight changes. And it could be often as the user types or scrolls through the code in a buffer.

### PR DESCRIPTION
**Pull request by yegappan**
Closes #8675.

### CODE REVIEW

**Comment by vim-ml** [comment]
Hi,

On Thu, Aug 12, 2021 at 10:58 PM Yegappan Lakshmanan <
***@***.***> wrote:

> Closes #8675 <https://github.com/vim/vim/issues/8675>.
> ------------------------------
> You can view, comment on, or merge this pull request online at:
>
>   https://github.com/vim/vim/pull/8751
> Commit Summary
>
>    - Add support for prop_add_list()
>
>
> The current implementation of the prop_add_list() function supports adding
a single text property to multiple locations in a buffer. Should we modify
this to support adding multiple text properties in a single call?

prop_add_list([[{prop1}, [<list of locations>]], [{prop2}, [<list of
locations>], ...])

- Yegappan

File: src/textprop.c [hunk 1]
```diff
@@ -351,8 +293,143 @@ prop_add_common(
...
+    linenr_T	end_lnum;
+    colnr_T	end_col;
+    char_u	*type_name;
+    buf_T	*buf = default_buf;
+    int		id = 0;
+
+    if (dict == NULL || dict_find(dict, (char_u *)"type", -1) == NULL)
+    {
+	emsg(_("E965: missing property type name"));
+	return;
```
Comment (dpelle) [inline src/textprop.c:401]: Indentation is wrong here.  It's also wrong in several other lines below (e.g. line 389)

File: src/testdir/test_textprop.vim [hunk 2]
```diff
@@ -339,6 +339,33 @@ func Test_prop_add()
...
+  call setline(1, ['one one one', 'two two two', 'six six six'])
+  call prop_add_list(#{type: 'one', id: 2}, [[1, 1, 3], [2, 5, 7], [3, 9, 11]])
+  call assert_equal([#{id: 2, col: 1, type_bufnr: 0, end: 1, type: 'one',
+        \ length: 2, start: 1}], prop_list(1))
+  call assert_equal([#{id: 2, col: 5, type_bufnr: 0, end: 1, type: 'one',
+        \ length: 2, start: 1}], prop_list(2))
+  call assert_equal([#{id: 2, col: 9, type_bufnr: 0, end: 1, type: 'one',
+        \ length: 2, start: 1}], prop_list(3))
+  call assert_fails('call prop_add_list([1, 2], [[1, 1, 3]])', 'E1206:')
+  call assert_fails('call prop_add_list({}, {})', 'E1211:')
```
Comment (dpelle) [inline src/testdir/test_textprop.vim:358]: Also test with:
* `test_null_list()`?
* a list that contains something else that a lists e.g. `[1, 'a']`

**Comment by bfrg** [comment]
I find it awkward that multiline text-properties cannot be handled with this function. Is there really such a drastic slowdown when the function is changed to `prop_add_list({props}, [[{lnum}, {col}, {end-lnum}, {end-col}], ...])`?

**Comment by vim-ml** [comment]
Hi,

Thanks for the comments. See below.

On Fri, Aug 13, 2021 at 12:37 AM Dominique Pellé ***@***.***>
wrote:

> ***@***.**** commented on this pull request.
> ------------------------------
>
> In src/textprop.c
> <https://github.com/vim/vim/pull/8751#discussion_r688304652>:
>
> > +	linenr_T    start_lnum,
> +	colnr_T	    start_col,
> +	dict_T	    *dict,
> +	buf_T	    *default_buf,
> +	typval_T    *dict_arg)
> +{
> +    linenr_T	end_lnum;
> +    colnr_T	end_col;
> +    char_u	*type_name;
> +    buf_T	*buf = default_buf;
> +    int		id = 0;
> +
> +    if (dict == NULL || dict_find(dict, (char_u *)"type", -1) == NULL)
> +    {
> +	emsg(_("E965: missing property type name"));
> +	return;
>
> Indentation is wrong here. It's also wrong in several other lines below
> (e.g. line 389)
>

I checked the indentation in the file and it looks correct to me.


> ------------------------------
>
> In src/testdir/test_textprop.vim
> <https://github.com/vim/vim/pull/8751#discussion_r688307163>:
>
> > @@ -339,6 +339,33 @@ func Test_prop_add()
>    bwipe!
>  endfunc
>
> +" Test for the prop_add_list() function
> +func Test_prop_add_list()
> +  new
> +  call AddPropTypes()
> +  call setline(1, ['one one one', 'two two two', 'six six six'])
> +  call prop_add_list(#{type: 'one', id: 2}, [[1, 1, 3], [2, 5, 7], [3, 9, 11]])
> +  call assert_equal([#{id: 2, col: 1, type_bufnr: 0, end: 1, type: 'one',
> +        \ length: 2, start: 1}], prop_list(1))
> +  call assert_equal([#{id: 2, col: 5, type_bufnr: 0, end: 1, type: 'one',
> +        \ length: 2, start: 1}], prop_list(2))
> +  call assert_equal([#{id: 2, col: 9, type_bufnr: 0, end: 1, type: 'one',
> +        \ length: 2, start: 1}], prop_list(3))
> +  call assert_fails('call prop_add_list([1, 2], [[1, 1, 3]])', 'E1206:')
> +  call assert_fails('call prop_add_list({}, {})', 'E1211:')
>
> Also test with:
>
>    - test_null_list()?
>
> I have added a new test for this.


>
>    - a list that contains something else that a lists e.g. [1, 'a']
>
>
> There is already a test for using a List with non-number items a few lines
below.

Regards,
Yegappan

**Comment by vim-ml** [comment]
Hi,

On Fri, Aug 13, 2021 at 5:18 AM bfrg ***@***.***> wrote:
>
> I find it awkward that multiline text-properties cannot be handled with this function.
> Is there really such a drastic slowdown when the function is changed to
> prop_add_list({props}, [[{lnum}, {col}, {end-lnum}, {end-col}], ...])?
>

Adding one more item to the List will not slow down the function.
The user now needs to pass a four item List instead of a three item
List. If the common case is single line text properties, then in most
use cases, lnum and end-lnum will be the same.

Based on Bram's comment, I thought if multi-line text property is needed,
then prop_add() should be used?

- Yegappan

**Comment by vim-ml** [comment]
> On Fri, Aug 13, 2021 at 5:18 AM bfrg ***@***.***> wrote:
> >
> > I find it awkward that multiline text-properties cannot be handled with this function.
> > Is there really such a drastic slowdown when the function is changed to
> > prop_add_list({props}, [[{lnum}, {col}, {end-lnum}, {end-col}], ...])?
> >
> 
> Adding one more item to the List will not slow down the function.
> The user now needs to pass a four item List instead of a three item
> List. If the common case is single line text properties, then in most
> use cases, lnum and end-lnum will be the same.
> 
> Based on Bram's comment, I thought if multi-line text property is needed,
> then prop_add() should be used?

Even though it's very likely that prop_add_list() will only be used for
single line items, and one can always fall back to calling prop_add()
multiple times, there is the argument of what users expect.  It's a bit
of a surprise if something is left out for unclear reasons.  Especially
when adding it doesn't really have a negative effect.

So we could make it [lnum, col, end_lnum, end_col].  Since this would be
filled in by a script, setting end_lnum equal to lnum won't be much
hassle, right?

-- 
Vi beats Emacs to death, and then again!
			http://linuxtoday.com/stories/5764.html

 /// Bram Moolenaar -- ***@***.*** -- http://www.Moolenaar.net   \\\
///                                                                      \\\
\\\        sponsor Vim, vote for features -- http://www.Vim.org/sponsor/ ///
 \\\            help me help AIDS victims -- http://ICCF-Holland.org    ///

**Comment by vim-ml** [comment]
Hi Bram,

On Fri, Aug 13, 2021 at 8:49 AM Bram Moolenaar ***@***.***> wrote:
>
>
> > On Fri, Aug 13, 2021 at 5:18 AM bfrg ***@***.***> wrote:
> > >
> > > I find it awkward that multiline text-properties cannot be handled with this function.
> > > Is there really such a drastic slowdown when the function is changed to
> > > prop_add_list({props}, [[{lnum}, {col}, {end-lnum}, {end-col}], ...])?
> > >
> >
> > Adding one more item to the List will not slow down the function.
> > The user now needs to pass a four item List instead of a three item
> > List. If the common case is single line text properties, then in most
> > use cases, lnum and end-lnum will be the same.
> >
> > Based on Bram's comment, I thought if multi-line text property is needed,
> > then prop_add() should be used?
>
> Even though it's very likely that prop_add_list() will only be used for
> single line items, and one can always fall back to calling prop_add()
> multiple times, there is the argument of what users expect.  It's a bit
> of a surprise if something is left out for unclear reasons.  Especially
> when adding it doesn't really have a negative effect.
>
> So we could make it [lnum, col, end_lnum, end_col].  Since this would be
> filled in by a script, setting end_lnum equal to lnum won't be much
> hassle, right?
>

I have updated the PR with the support for specifying the end line number.

Should we also support specifying multiple text properties in a single call?

prop_add_list([[{prop1}, [<locations]], [{prop2}, [<locations>]], ....])

Regards,
Yegappan

**Comment by vim-ml** [comment]
Yegappan wrote:

> On Fri, Aug 13, 2021 at 8:49 AM Bram Moolenaar ***@***.***> wrote:
> >
> >
> > > On Fri, Aug 13, 2021 at 5:18 AM bfrg ***@***.***> wrote:
> > > >
> > > > I find it awkward that multiline text-properties cannot be handled with this function.
> > > > Is there really such a drastic slowdown when the function is changed to
> > > > prop_add_list({props}, [[{lnum}, {col}, {end-lnum}, {end-col}], ...])?
> > > >
> > >
> > > Adding one more item to the List will not slow down the function.
> > > The user now needs to pass a four item List instead of a three item
> > > List. If the common case is single line text properties, then in most
> > > use cases, lnum and end-lnum will be the same.
> > >
> > > Based on Bram's comment, I thought if multi-line text property is needed,
> > > then prop_add() should be used?
> >
> > Even though it's very likely that prop_add_list() will only be used for
> > single line items, and one can always fall back to calling prop_add()
> > multiple times, there is the argument of what users expect.  It's a bit
> > of a surprise if something is left out for unclear reasons.  Especially
> > when adding it doesn't really have a negative effect.
> >
> > So we could make it [lnum, col, end_lnum, end_col].  Since this would be
> > filled in by a script, setting end_lnum equal to lnum won't be much
> > hassle, right?
> >
> 
> I have updated the PR with the support for specifying the end line number.
> 
> Should we also support specifying multiple text properties in a single call?
> 
> prop_add_list([[{prop1}, [<locations]], [{prop2}, [<locations>]], ....])

That adds more dictionaries, which is less efficient and results in a
mix of types, which is more complicated.  I think we should stick to
using one {props} argument that applies to the list of locations.
If we really need more, we could add another function.

-- 
TIM:    Too late.
ARTHUR: What?
TIM:    There he is!
   [They all turn, and see a large white RABBIT lollop a few yards out of the
   cave.  Accompanied by terrifying chord and jarring metallic monster noise.]
                 "Monty Python and the Holy Grail" PYTHON (MONTY) PICTURES LTD

 /// Bram Moolenaar -- ***@***.*** -- http://www.Moolenaar.net   \\\
///                                                                      \\\
\\\        sponsor Vim, vote for features -- http://www.Vim.org/sponsor/ ///
 \\\            help me help AIDS victims -- http://ICCF-Holland.org    ///

**Comment by vim-ml** [comment]
Hi Bram,

On Fri, Aug 13, 2021 at 10:46 AM Bram Moolenaar ***@***.***> wrote:
>
>
> Yegappan wrote:
>
> > On Fri, Aug 13, 2021 at 8:49 AM Bram Moolenaar ***@***.***> wrote:
> > >
> > >
> > > > On Fri, Aug 13, 2021 at 5:18 AM bfrg ***@***.***> wrote:
> > > > >
> > > > > I find it awkward that multiline text-properties cannot be handled with this function.
> > > > > Is there really such a drastic slowdown when the function is changed to
> > > > > prop_add_list({props}, [[{lnum}, {col}, {end-lnum}, {end-col}], ...])?
> > > > >
> > > >
> > > > Adding one more item to the List will not slow down the function.
> > > > The user now needs to pass a four item List instead of a three item
> > > > List. If the common case is single line text properties, then in most
> > > > use cases, lnum and end-lnum will be the same.
> > > >
> > > > Based on Bram's comment, I thought if multi-line text property is needed,
> > > > then prop_add() should be used?
> > >
> > > Even though it's very likely that prop_add_list() will only be used for
> > > single line items, and one can always fall back to calling prop_add()
> > > multiple times, there is the argument of what users expect.  It's a bit
> > > of a surprise if something is left out for unclear reasons.  Especially
> > > when adding it doesn't really have a negative effect.
> > >
> > > So we could make it [lnum, col, end_lnum, end_col].  Since this would be
> > > filled in by a script, setting end_lnum equal to lnum won't be much
> > > hassle, right?
> > >
> >
> > I have updated the PR with the support for specifying the end line number.
> >
> > Should we also support specifying multiple text properties in a single call?
> >
> > prop_add_list([[{prop1}, [<locations]], [{prop2}, [<locations>]], ....])
>
> That adds more dictionaries, which is less efficient and results in a
> mix of types, which is more complicated.  I think we should stick to
> using one {props} argument that applies to the list of locations.
> If we really need more, we could add another function.
>

I think I didn't provide a clear description. I meant, to add multiple
text properties in several locations, instead of calling the
prop_add_list() function multiple times (one for each text property
and a list of locations), we can change the prop_add_list() function
to accept a single List. In this List, each item is a List with the text
property {prop} and a list of locations. This can help in reducing
the number of calls to prop_add_list().

Regards,
Yegappan

**Comment by vim-ml** [comment]
Yegappan wrote:

> > > On Fri, Aug 13, 2021 at 8:49 AM Bram Moolenaar ***@***.***> wrote:
> > > >
> > > >
> > > > > On Fri, Aug 13, 2021 at 5:18 AM bfrg ***@***.***> wrote:
> > > > > >
> > > > > > I find it awkward that multiline text-properties cannot be handled with this function.
> > > > > > Is there really such a drastic slowdown when the function is changed to
> > > > > > prop_add_list({props}, [[{lnum}, {col}, {end-lnum}, {end-col}], ...])?
> > > > > >
> > > > >
> > > > > Adding one more item to the List will not slow down the function.
> > > > > The user now needs to pass a four item List instead of a three item
> > > > > List. If the common case is single line text properties, then in most
> > > > > use cases, lnum and end-lnum will be the same.
> > > > >
> > > > > Based on Bram's comment, I thought if multi-line text property is needed,
> > > > > then prop_add() should be used?
> > > >
> > > > Even though it's very likely that prop_add_list() will only be used for
> > > > single line items, and one can always fall back to calling prop_add()
> > > > multiple times, there is the argument of what users expect.  It's a bit
> > > > of a surprise if something is left out for unclear reasons.  Especially
> > > > when adding it doesn't really have a negative effect.
> > > >
> > > > So we could make it [lnum, col, end_lnum, end_col].  Since this would be
> > > > filled in by a script, setting end_lnum equal to lnum won't be much
> > > > hassle, right?
> > > >
> > >
> > > I have updated the PR with the support for specifying the end line number.
> > >
> > > Should we also support specifying multiple text properties in a single call?
> > >
> > > prop_add_list([[{prop1}, [<locations]], [{prop2}, [<locations>]], ....])
> >
> > That adds more dictionaries, which is less efficient and results in a
> > mix of types, which is more complicated.  I think we should stick to
> > using one {props} argument that applies to the list of locations.
> > If we really need more, we could add another function.
> >
> 
> I think I didn't provide a clear description. I meant, to add multiple
> text properties in several locations, instead of calling the
> prop_add_list() function multiple times (one for each text property
> and a list of locations), we can change the prop_add_list() function
> to accept a single List. In this List, each item is a List with the text
> property {prop} and a list of locations. This can help in reducing
> the number of calls to prop_add_list().

I understand.  This is starting to look a lot like a sequence of calls
to prop_add().  That takes away lot of the "added value" of the
function.  The idea is to add properties with the same attributes in a
list of positions, thus avoiding to list the properties multiple times.

Potential users of this function: Please comment.

-- 
If "R" is Reverse, how come "D" is FORWARD?

 /// Bram Moolenaar -- ***@***.*** -- http://www.Moolenaar.net   \\\
///                                                                      \\\
\\\        sponsor Vim, vote for features -- http://www.Vim.org/sponsor/ ///
 \\\            help me help AIDS victims -- http://ICCF-Holland.org    ///

**Comment by brammool** [comment]
Since there are no recent comments I'll include this now.

**Pull request closed without merging**